warehouse/slave.go

```diff
@@ -101,40 +119,58 @@ func (job *PayloadT) getFileManager() (filemanager.FileManager, error) {
 
 /*
  * Download Staging file for the job
+ * If error occurs with the current config and current revision is different from staging revision
+ * We retry with the staging revision config if it is present
  */
 func (jobRun *JobRunT) downloadStagingFile() error {
-	filePath := jobRun.stagingFilePath
-	file, err := os.Create(filePath)
-	if err != nil {
-		panic(err)
-	}
-
 	job := jobRun.job
-	downloader, err := jobRun.job.getFileManager()
-	if err != nil {
-		pkgLogger.Errorf("[WH]: Failed to initialize downloader")
-		return err
-	}
+	downloadTask := func(config interface{}, useRudderStorage bool) (err error) {
+		filePath := jobRun.stagingFilePath
+		file, err := os.Create(filePath)
+		if err != nil {
+			panic(err)
+		}
 
-	timer := jobRun.timerStat("download_staging_file_time")
-	timer.Start()
+		downloader, err := job.getFileManager(config, useRudderStorage)
+		if err != nil {
+			pkgLogger.Errorf("[WH]: Failed to initialize downloader")
+			return err
+		}
 
-	err = downloader.Download(context.TODO(), file, job.StagingFileLocation)
-	if err != nil {
-		pkgLogger.Errorf("[WH]: Failed to download file")
-		return err
+		timer := jobRun.timerStat("download_staging_file_time")
+		timer.Start()
+
+		err = downloader.Download(context.TODO(), file, job.StagingFileLocation)
+		if err != nil {
+			pkgLogger.Errorf("[WH]: Failed to download file")
+			return err
+		}
+		file.Close()
+		timer.End()
+
+		fi, err := os.Stat(filePath)
+		if err != nil {
+			pkgLogger.Errorf("[WH]: Error getting file size of downloaded staging file: ", err)
+			return err
+		}
+		fileSize := fi.Size()
+		pkgLogger.Debugf("[WH]: Downloaded staging file %s size:%v", job.StagingFileLocation, fileSize)
+		return
 	}
-	file.Close()
-	timer.End()
 
-	fi, err := os.Stat(filePath)
+	err := downloadTask(job.CurrentDestinationConfig, job.CurrentUseRudderStorage)
 	if err != nil {
-		pkgLogger.Errorf("[WH]: Error getting file size of downloaded staging file: ", err)
-		return err
+		if job.StagingDestinationRevisionID != job.CurrentDestinationRevisionID && job.StagingDestinationConfig != nil {
```

Comment:
I would wrap this up in a condition, to allow for better comprehension.





```suggestion
	if ( pickupStagingFileBucket(job) ) {
            err = downloadTask()
}
```

Reply:
Wdyt ?

Reply:
Makes sense. Will add test for that as well.